<a href="https://colab.research.google.com/github/Ckiteme/CKiteme-Assignment-Model-Quality-and-Improvements/blob/main/CKiteme_Assignment_Model_Quality_and_Improvements.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Instructions**

*Problem Statement*

As a data professional working for a pharmaceutical company, you need to develop a
model that predicts whether a patient will be diagnosed with diabetes. The model needs
to have an accuracy score greater than 0.85.
You will be required to document the following steps:
● Data Importation
● Data Exploration
● Data Cleaning
● Data Preparation
● Data Modeling (Using Decision Trees, Random Forest and Logistic Regression)
● Model Evaluation
● Hyparameter Tuning
● Findings and Recommendations
Dataset
● Dataset URL: https://bit.ly/DiabetesDS

In [1]:
#importing the libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics import classification_report,confusion_matrix

**1. Data Importation**

In [2]:
# Your code goes here
# ---
#
import pandas as pd
import numpy as np
diabetes = pd.read_csv('https://bit.ly/DiabetesDS')
diabetes.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

**2. Data Exploration, Data Cleaning and Data Preparation.**

In [3]:
# preview dataset shape
diabetes.shape

(768, 9)

In [4]:
# look for duplicates

diabetes.duplicated().sum()

0

In [5]:
# look for missing records
diabetes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [6]:
# look for missing records

diabetes.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [7]:
# preview datatypes

diabetes.dtypes

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [8]:
diabetes.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [9]:
diabetes['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

**3.Data Modeling**

In [12]:
#create features and target variables & spliting the dataset
features= diabetes.drop('Outcome',axis=1)

target = diabetes['Outcome']

features_train, features_test, target_train, target_test= train_test_split(

    features, target, test_size=0.20, random_state=12345
)

In [13]:
#Predicting the outcome based on test dataset and storing in the variable
model = RandomForestClassifier(random_state=12345)
        
model.fit(features_train,target_train)

predictions = model.predict(features_test)

accuracy_score(predictions, target_test)

0.8181818181818182

In [14]:
model = DecisionTreeClassifier(random_state=12345)
        
model.fit(features_train,target_train)

predictions = model.predict(features_test)

accuracy_score(predictions, target_test)

0.7792207792207793

In [15]:
model = LogisticRegression(random_state=12345)
        
model.fit(features_train,target_train)

predictions = model.predict(features_test)

accuracy_score(predictions, target_test)

0.7922077922077922

In [16]:
predictions = model.predict(features_test)
accuracy_score(target_test, predictions)

0.7922077922077922

**4. Model Evaluation**

In [17]:
print("Confusion Matrix : \n",confusion_matrix( target_test , predictions))

Confusion Matrix : 
 [[100   6]
 [ 26  22]]


In [18]:
for depth in range(1, 6):
        model = DecisionTreeClassifier(max_depth=depth)

        model.fit(features_test, target_test)

        predictions = model.predict(features_test)
        accuracy_score(target_test, predictions)

        print("max_depth =", depth, ": ", end='')
        print(accuracy_score(target_test, predictions))

max_depth = 1 : 0.7532467532467533
max_depth = 2 : 0.7987012987012987
max_depth = 3 : 0.8311688311688312
max_depth = 4 : 0.8961038961038961
max_depth = 5 : 0.948051948051948


**5. Hyparameter Tuning**

In [19]:
for depth in range(1, 6):
        model =  DecisionTreeClassifier(random_state=12345,max_depth=depth)

        model.fit(features_train, target_train)

       
        predictions = model.predict(features_test)

        print("max_depth =", depth, ": ", end='')
        print(accuracy_score(target_test, predictions))

max_depth = 1 : 0.7597402597402597
max_depth = 2 : 0.7922077922077922
max_depth = 3 : 0.7857142857142857
max_depth = 4 : 0.7922077922077922
max_depth = 5 : 0.7662337662337663


**6. Findings and Recommendations**

After analysing the data with various Prediction Models, I have found that RandomForestClassifier was the best model with an accuracy score of 82%